### ISSUE TITLE
Boost allocation performance with MS snmalloc

### ISSUE THREAD

**Issue by WSLUser** [contour-terminal/contour#397]
Use https://github.com/microsoft/snmalloc, which works on Windows, Linux and BSD (so technically Mac should be supported too) to boost the perfomance of allocations.
## Motivation

I thought it would be useful to have a small C++ header library that officially comes from MS used for boosting allocation performance.

**Comment by christianparpart** [comment]
There is [tcmalloc](https://github.com/google/tcmalloc) I am specifically interested, a malloc implementation by Google. With snmalloc now on the table, I wonder if there isn't some kind of malloc implementation performance analysis has been already made by others. I think generally speaking it should be possible to provide a CMake flag that can compile-time switch between all those malloc implementations, such as: native, jemalloc, tcmalloc, snmalloc, mimalloc, ... anything else?

C++ has great support for custom allocators, something I do not make really use of currently. It might be worth investigating in whether or not such a thing would make sense, too. OTOH, I am currently redesigning the grid system to live without any dynamic allocations (or: reduce them as much as possible and keep the memory regions dense)

**Comment by christianparpart** [comment]
Gotcha: https://github.com/microsoft/snmalloc/blob/master/snmalloc.pdf

**Comment by WSLUser** [comment]
Yes nice benchmark find. Near the bottom of that pdf and it's likely the project evolved further from the time of that benchmark. Even there this one appears to be the best one available. I am not surprised Google doesn't do as well as MS. They never do.

**Comment by skyline75489** [comment]
There's more of them: https://github.com/microsoft/mimalloc :)

**Comment by WSLUser** [comment]
So MS has two different ones? They're both being actively maintained and they don't show comparisons between them. Are 2 different teams working on essentially the same type of thing? From what I can tell, snmalloc still appears to be preferable as it's a small header library and mimalloc is not but the benchmark would need to be used to compare against them to see which would be worth it more to implement.

**Comment by data-man** [comment]
#341 ;)

> they don't show comparisons between them

https://github.com/mjansson/rpmalloc-benchmark

**Comment by WSLUser** [comment]
That doesn't show snmalloc vs mimalloc. but it is yet another benchmark posted. I suspect they'll all have benchmark comparisons only against the ones that aren't as good as them of course.

**Comment by christianparpart** [comment]
@WSLUser I think they both serve different usecases. snmalloc was specifically designed for good performance in cross malloc/free calls (malloc in one, free in other threads). that usecase we do not have in Contour as we have two main threads (terminal I/O and GUI rendering thread). they both communicate but the important malloc's have their frees on the terminal I/O side (fetching the renderable cells). Performance is however much more of a concern within the terminal I/O thread itself. and that can be seen as single threaded, so snmalloc (as much as header-only libraries are hyped) does not really give a huge gain as far as that part is concerned. mimalloc we already have in though. I'm not strictly against a compile time flag for others as long as the maintenance burden is low)

**Comment by jerch** [comment]
Hmm idk, is caring about a different general purpose allocator not way too much micro-optimization atm? From my perspective a terminal has very limited allocation/free needs, if the internal buffer layout is chosen carefully. This includes borrowing as much as possible (less allocation + copy), and maybe do some own pooling. For the latter thinking about memory management is indeed important, but often the problem turns out to be very specific and can be solved by a much faster own pool allocator.
